File: app/src/com/ternaryop/phototumblrshare/dialogs/SchedulePostDialog.java
package com.ternaryop.phototumblrshare.dialogs;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

import org.json.JSONObject;

import android.app.DatePickerDialog;
import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.Context;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.TextView;
import android.widget.TimePicker;

import com.ternaryop.phototumblrshare.R;
import com.ternaryop.phototumblrshare.list.PhotoSharePost;
import com.ternaryop.tumblr.AbsCallback;
import com.ternaryop.tumblr.Tumblr;

public class SchedulePostDialog extends Dialog implements View.OnClickListener, TimePickerDialog.OnTimeSetListener, DatePickerDialog.OnDateSetListener {

	private final Calendar scheduleDateTime;
	private Button chooseDateButton;
	private Button chooseTimeButton;
	private SimpleDateFormat timeFormat;
	private SimpleDateFormat dateFormat;
	private final PhotoSharePost item;
	private final String blogName;
	private final onPostScheduleListener onPostSchedule;

	public SchedulePostDialog(Context context, String blogName, PhotoSharePost item, Calendar scheduleDateTime, onPostScheduleListener onPostSchedule) {
		super(context);
		this.blogName = blogName;
		this.item = item;
		this.scheduleDateTime = scheduleDateTime;
		this.onPostSchedule = onPostSchedule;
		setContentView(R.layout.dialog_schedule_post);
		
		setTitle(R.string.schedule_post);
		((TextView)findViewById(R.id.post_title_textview)).setText(item.getFirstTag());
		((Button)findViewById(R.id.cancelButton)).setOnClickListener(this);
		((Button)findViewById(R.id.schedule_button)).setOnClickListener(this);

		// init date
		chooseDateButton = (Button)findViewById(R.id.choose_date_button);
		dateFormat = new SimpleDateFormat("dd/MM/yyyy", Locale.US);
		chooseDateButton.setText(dateFormat.format(scheduleDateTime.getTime()));
		chooseDateButton.setOnClickListener(this);

		// init time
		chooseTimeButton = (Button)findViewById(R.id.choose_time_button);
		timeFormat = new SimpleDateFormat("HH:mm", Locale.US);
		chooseTimeButton.setText(timeFormat.format(scheduleDateTime.getTime()));
		chooseTimeButton.setOnClickListener(this);
	}

	@Override
	public void onClick(View v) {
		switch (v.getId()) {
			case R.id.cancelButton:
				dismiss();
				return;
			case R.id.schedule_button:
				schedulePost();
				return;
			case R.id.choose_time_button:
				new TimePickerDialog(getContext(), this,
						scheduleDateTime.get(Calendar.HOUR_OF_DAY),
						scheduleDateTime.get(Calendar.MINUTE), true)
				.show();
				return;
			case R.id.choose_date_button:
				new DatePickerDialog(getContext(), this,
						scheduleDateTime.get(Calendar.YEAR),
						scheduleDateTime.get(Calendar.MONTH),
						scheduleDateTime.get(Calendar.DAY_OF_MONTH))
				.show();
				return;
		}
	}

	private void schedulePost() {
		Tumblr.getSharedTumblr(getContext())
		.schedulePost(blogName, item.getPostId(), scheduleDateTime.getTimeInMillis(), new AbsCallback(this, R.string.parsing_error) {

			@Override
			public void complete(JSONObject result) {
				if (onPostSchedule != null) {
					onPostSchedule.onPostScheduled(item.getPostId(), scheduleDateTime);
				}
				dismiss();
			}
		});
	}

	public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
		scheduleDateTime.set(Calendar.HOUR_OF_DAY, hourOfDay);
		scheduleDateTime.set(Calendar.MINUTE, minute);
		chooseTimeButton.setText(timeFormat.format(scheduleDateTime.getTime()));
	}

	@Override
	public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
		scheduleDateTime.set(Calendar.YEAR, year);
		scheduleDateTime.set(Calendar.MONTH, monthOfYear);
		scheduleDateTime.set(Calendar.DAY_OF_MONTH, dayOfMonth);
		chooseDateButton.setText(dateFormat.format(scheduleDateTime.getTime()));
	}
	
	public interface onPostScheduleListener {
		void onPostScheduled(long id, Calendar scheduledDateTime);
	}
}

File: tumblr-dialog/src/main/java/com/ternaryop/photoshelf/tumblr/dialog/SchedulePostDialog.kt
package com.ternaryop.photoshelf.tumblr.dialog

import android.annotation.SuppressLint
import android.app.DatePickerDialog
import android.app.Dialog
import android.app.TimePickerDialog
import android.os.Bundle
import android.text.format.DateUtils
import android.view.View
import android.widget.Button
import android.widget.DatePicker
import android.widget.TimePicker
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import com.ternaryop.tumblr.TumblrPost
import com.ternaryop.utils.date.dayOfMonth
import com.ternaryop.utils.date.hourOfDay
import com.ternaryop.utils.date.minute
import com.ternaryop.utils.date.month
import com.ternaryop.utils.date.year
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

class SchedulePostDialog : DialogFragment(), View.OnClickListener,
    TimePickerDialog.OnTimeSetListener, DatePickerDialog.OnDateSetListener {

    private lateinit var post: TumblrPost
    private lateinit var scheduleDateTime: Calendar
    private lateinit var chooseDateButton: Button
    private lateinit var chooseTimeButton: Button
    private lateinit var timeFormat: SimpleDateFormat
    private lateinit var scheduleButton: Button

    @SuppressLint("InflateParams") // for dialogs passing null for root is valid, ignore the warning
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val view = requireActivity().layoutInflater.inflate(R.layout.dialog_schedule_post, null)

        post = checkNotNull(arguments?.getSerializable(ARG_POST) as? TumblrPost)
        scheduleDateTime = checkNotNull(arguments?.getSerializable(ARG_SCHEDULE_DATE) as? Calendar)

        setupUI(view)

        val dialog = AlertDialog.Builder(requireContext())
            .setView(view)
            .setNegativeButton(android.R.string.cancel) { _, _ -> dismiss() }
            .setPositiveButton(R.string.schedule_title) { _, _ -> }
            .create()

        dialog.setOnShowListener {
            scheduleButton = dialog.getButton(AlertDialog.BUTTON_POSITIVE)
            scheduleButton.setOnClickListener {
                checkAndSchedule()
            }
        }
        return dialog
    }

    private fun setupUI(view: View) {
        // init date
        chooseDateButton = view.findViewById(R.id.choose_date_button)
        chooseDateButton.setOnClickListener(this)
        updateDateButton()

        // init time
        chooseTimeButton = view.findViewById(R.id.choose_time_button)
        timeFormat = SimpleDateFormat("HH:mm", Locale.US)
        updateTimeButton(scheduleDateTime)
        chooseTimeButton.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.choose_time_button -> TimePickerDialog(context, this,
                scheduleDateTime.hourOfDay,
                scheduleDateTime.minute, true)
                .show()
            R.id.choose_date_button -> {
                val dialog = DatePickerDialog(requireContext(), this,
                    scheduleDateTime.year,
                    scheduleDateTime.month,
                    scheduleDateTime.dayOfMonth)
                dialog.datePicker.minDate = System.currentTimeMillis()
                dialog.show()
            }
        }
    }

    private fun checkAndSchedule() {
        if (scheduleDateTime.timeInMillis <= System.currentTimeMillis()) {
            AlertDialog.Builder(requireContext())
                .setMessage(R.string.scheduled_time_is_in_the_past_continue_title)
                .setPositiveButton(android.R.string.ok) { _, _ -> schedulePost() }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
        } else {
            schedulePost()
        }
    }

    private fun schedulePost() {
        scheduleButton.isEnabled = false
        parentFragmentManager.setFragmentResult(
            checkNotNull(arguments?.getString(EXTRA_REQUEST_KEY)),
            Bundle().also {
                parentFragmentManager.putFragment(it, EXTRA_DIALOG, this)
                it.putSerializable(EXTRA_SCHEDULE_DATA, SchedulePostData(post, scheduleDateTime))
            }
        )
    }

    override fun onTimeSet(view: TimePicker, hourOfDay: Int, minute: Int) {
        scheduleDateTime.hourOfDay = hourOfDay
        scheduleDateTime.minute = minute
        updateTimeButton(scheduleDateTime)
    }

    override fun onDateSet(view: DatePicker, year: Int, monthOfYear: Int, dayOfMonth: Int) {
        scheduleDateTime.year = year
        scheduleDateTime.month = monthOfYear
        scheduleDateTime.dayOfMonth = dayOfMonth
        updateDateButton()
    }

    private fun updateDateButton() {
        val timeString = DateUtils.getRelativeTimeSpanString(
            scheduleDateTime.timeInMillis,
            System.currentTimeMillis(),
            DateUtils.DAY_IN_MILLIS,
            DateUtils.FORMAT_SHOW_DATE)

        chooseDateButton.text = timeString
    }

    private fun updateTimeButton(scheduleDateTime: Calendar) {
        chooseTimeButton.text = timeFormat.format(scheduleDateTime.time)
    }

    companion object {
        private const val ARG_SCHEDULE_DATE = "scheduleDate"
        private const val ARG_POST = "post"
        private const val EXTRA_REQUEST_KEY = "requestKey"
        const val EXTRA_DIALOG = "dialog"
        const val EXTRA_SCHEDULE_DATA = "scheduleData"

        fun newInstance(
            post: TumblrPost,
            scheduleDateTime: Calendar,
            requestKey: String
        ) = SchedulePostDialog().apply {
            arguments = bundleOf(
                EXTRA_REQUEST_KEY to requestKey,
                ARG_SCHEDULE_DATE to scheduleDateTime,
                ARG_POST to post
            )
        }
    }
}
